File: Benchmarks/PageRank/C#/PageRankRandom.cs
class GraphRandom
{
    public System.Collections.Generic.List<int>[] Outgoing;
    public System.Collections.Generic.List<int>[] Incoming;
    public int NVertices = 0;

    public GraphRandom(int vertices, float density)
    {
        NVertices = vertices;
        Outgoing = new System.Collections.Generic.List<int>[vertices];
        Incoming = new System.Collections.Generic.List<int>[vertices];

        for (int i = 0; i < vertices; i++)
        {
            Outgoing[i] = new System.Collections.Generic.List<int>();
            Incoming[i] = new System.Collections.Generic.List<int>();
        }

        System.Random random = new System.Random();
        for (int i = 0; i < vertices; i++)
        {
            for (int j = 0; j < vertices; j++)
            {
                float vertex = (float)random.NextDouble();
                if (i != j && vertex < density)
                {
                    Outgoing[i].Add(j);
                    Incoming[j].Add(i);
                }
            }
        }
    }
}

class PageRankRandom
{
    public static float PageRank_single(int v, GraphRandom graph, float[] ranks, float damping)
    {
        float rank = (1.0f - damping) / graph.NVertices;

        foreach (int u in graph.Incoming[v])
            rank += damping * ranks[u] / graph.Outgoing[u].Count;

        return rank;
    }

    public static float[] PageRank(GraphRandom graph, float damping, float epsilon = 1e-4f)
    {
        float[] ranks = new float[graph.NVertices];
        float[] newRanks = new float[graph.NVertices];

        for (int v = 0; v < graph.NVertices; v++)
        {
            ranks[v] = 1f / graph.NVertices;
            newRanks[v] = 1f / graph.NVertices;
        }


        float change = epsilon + 1f;
        while (change > epsilon)
        {

            for (int v = 0; v < graph.NVertices; v++)
                newRanks[v] = PageRank_single(v, graph, ranks, damping);

            // Recalculate change
            change = 0f;
            for (int v = 0; v < graph.NVertices; v++)
                change += System.MathF.Abs(ranks[v] - newRanks[v]);

            for (int v = 0; v < graph.NVertices; v++)
                ranks[v] = newRanks[v];
        }

        return ranks;
    }

    public static void Main(string[] args)
    {
        int vertices = int.Parse(args[0]);
        float density = float.Parse(args[1]);
        int iterations = int.Parse(args[2]);

        var graph = new GraphRandom(vertices, density);

        float totalSum = 0.0f;

        for (int i = 0; i < iterations; i++)
        {
            var ranks = PageRank(graph, 0.85f + i / 1000.0f); 
            for (int r = 0; r < vertices; r++)
            {
                totalSum += ranks[r];
            }
        }

        System.Console.WriteLine(totalSum);
    }
}

File: Benchmarks/PageRank/Java/PageRankRandom.java
import java.util.ArrayList;
import java.util.Random;

class GraphRandom {

  public int nvertices;
  public ArrayList<ArrayList<Integer>> incoming;
  public ArrayList<ArrayList<Integer>> outgoing;

  public GraphRandom(int vertices, float density) {
    this.nvertices = vertices;
    this.incoming = new ArrayList<>();
    this.outgoing = new ArrayList<>();

    for (int i = 0; i < vertices; i++) {
      this.incoming.add(new ArrayList<>());
      this.outgoing.add(new ArrayList<>());
    }

    Random random = new Random();
    for (int i = 0; i < vertices; i++) {
      for (int j = 0; j < vertices; j++) {
        float vertex = random.nextFloat();
        if (i != j && vertex < density) {
          this.incoming.get(i).add(j);
          this.outgoing.get(j).add(i);
        }
      }
    }
  }
}

class PageRankRandom {

  public static float PageRank_single(int v, GraphRandom graph, float[] ranks,
                                      float damping) {
    float rank = (1f - damping) / (float)graph.nvertices;
    for (int u : graph.incoming.get(v)) {
      rank += damping * ranks[u] / (float)graph.outgoing.get(u).size();
    }

    return rank;
  }

  public static float[] PageRank(GraphRandom graph, float damping,
                                 float epsilon) {
    float[] ranks = new float[graph.nvertices];
    float[] newRanks = new float[graph.nvertices];

    for (int v = 0; v < graph.nvertices; v++) {
      ranks[v] = 1f / (float)graph.nvertices;
      newRanks[v] = 1f / (float)graph.nvertices;
    }

    float change = epsilon + 1f;
    while (change > epsilon) {
      for (int v = 0; v < graph.nvertices; v++) {
        newRanks[v] = PageRank_single(v, graph, ranks, damping);
      }

      change = 0f;
      for (int v = 0; v < graph.nvertices; v++) {
        change += Math.abs(ranks[v] - newRanks[v]);
      }

      for (int v = 0; v < graph.nvertices; v++) {
        ranks[v] = newRanks[v];
      }
    }

    return ranks;
  }

  public static void main(String[] args) {
    int vertices = Integer.parseInt(args[0]);
    float density = Float.parseFloat(args[1]);
    int iterations = Integer.parseInt(args[2]);

    GraphRandom graph = new GraphRandom(vertices, density);
    float totalSum = 0.0f;

    for (int i = 0; i < iterations; i++) {
      float[] ranks = PageRank(graph, 0.85f + (float)i / 1000.0f, 1e-4f);
      for (int r = 0; r < vertices; r++) {
        totalSum += ranks[r];
      }
    }

    System.out.println(totalSum);
  }
}
